node/src/tasks/user_submitted_file.rs

```diff
@@ -0,0 +1,111 @@
+        Self {
+            storage_hub_handler: self.storage_hub_handler.clone(),
+        }
+    }
+}
+
+impl<SHC: StorageHubHandlerConfig> UserSubmittedFileTask<SHC> {
+    pub fn new(storage_hub_handler: StorageHubHandler<SHC>) -> Self {
+        Self {
+            storage_hub_handler,
+        }
+    }
+}
+
+impl<SHC: StorageHubHandlerConfig> EventHandler<AcceptedBspVolunteer>
+    for UserSubmittedFileTask<SHC>
+{
+    async fn handle_event(&self, event: AcceptedBspVolunteer) -> anyhow::Result<()> {
+        info!(
+            target: LOG_TARGET,
+            "Handling file submitted by user {:?} to BSP with location {:?}",
+            event.owner,
+            event.location,
+        );
+
+        // TODO: use `Multiaddr` instead of `String`.
+        let multiaddresses = event.multiaddresses;
+        let file_metadata = Metadata {
+            owner: event.owner.to_string(),
+            size: event.size.into(),
+            // TODO: use `FileLocation` instead of `String`.
+            location: format!("{:?}", event.location.into_inner()),
+            fingerprint: event.fingerprint,
+        };
+        let chunk_count = file_metadata.chunk_count();
+
+        // TODO(Arthur): double check this, I'm assuming Blake2 as the Trie hash function.
+        let file_key = Blake2Hasher::hash(&serde_json::to_vec(&file_metadata)?.to_owned());
+
+        let peer_ids = multiaddresses
+            .iter()
+            .map(|multiaddr| {
+                let multiaddr_str = multiaddr
+                    .split("/")
+                    .last()
+                    .expect("Multiaddress of Bsp should not be empty.");
+                let multiaddr =
+                    Multiaddr::from_str(multiaddr_str).expect("Failed to convert into Multiaddr");
+                PeerId::try_from_multiaddr(&multiaddr).expect("Multiaddr without PeerId")
+            })
+            .collect::<Vec<PeerId>>();
+
+        for peer_id in peer_ids {
+            for chunk_id in 0..chunk_count {
+                // Acquire the write lock for the file storage.
+                let file_storage_lock: tokio::sync::RwLockWriteGuard<
+                    <SHC as StorageHubHandlerConfig>::FileStorage,
+                > = self.storage_hub_handler.file_storage.write().await;
+
+                let proof = file_storage_lock
+                    .generate_proof(&file_key, &chunk_id)
+                    .expect("File is not in storage, or proof does not exist.");
+
+                // Drop write lock after getting proof.
+                drop(file_storage_lock);
+
+                let upload_response = self
+                    .storage_hub_handler
+                    .file_transfer
+                    .upload_request(peer_id, file_key, proof)
+                    .await;
+
+                match upload_response {
+                    Ok(_) => {
+                        info!(target: LOG_TARGET, "Successfully uploaded chunk id {:?} to peer {:?}", chunk_id, peer_id)
```

Comment:
I would make this debug! since there would be quite a lot of chunks to send 😅

Reply:
Yes, good idea.